src/App.tsx

```diff
@@ -1,24 +1,249 @@
+  const [error, setError] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+
+  const newTodoFieldRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (newTodoFieldRef.current) {
+      newTodoFieldRef.current.focus();
+    }
+  }, [isLoading]);
+
+  useEffect(() => {
+    setError('');
+
+    getTodos()
+      .then(setTodos)
+      .catch(() => {
+        setError(errorMessages.load_todos);
+      });
+  }, []);
+
+  const timerIdRef = useRef<number>(0);
+
+  useEffect(() => {
+    if (timerIdRef.current) {
+      window.clearTimeout(timerIdRef.current);
+    }
+
+    timerIdRef.current = window.setTimeout(() => {
+      setError('');
+    }, 3000);
+  }, [error]);
+
+  const handleAddTodo = (todoTitle: string) => {
+    setTempTodo({
+      id: 0,
+      title: todoTitle.trim(),
+      userId: 0,
+      completed: false,
+    });
+
+    setIsLoading(true);
+
+    return addTodo(todoTitle)
+      .then((newTodo) => {
+        setTodos((prevTodos) => [...prevTodos, newTodo]);
+
+        newTodoFieldRef.current?.focus();
+      })
+      .catch(() => {
+        setError(errorMessages.add_todo);
+        throw new Error(errorMessages.add_todo);
+      })
+      .finally(() => {
+        setTempTodo(null);
+        setIsLoading(false);
+      });
+  };
+
+  const handleDeleteTodo = (todoId: number) => {
+    setLoadingTodoIds((prevTodoIds) => [...prevTodoIds, todoId]);
+    setIsLoading(true);
+
+    deleteTodo(todoId)
+      .then(() => {
+        setTodos((prevTodos) => prevTodos.filter((todo) => todo.id !== todoId));
+
+        newTodoFieldRef.current?.focus();
+      })
+      .catch(() => {
+        setError(errorMessages.delete_todo);
+      })
+      .finally(() => {
+        setLoadingTodoIds((prevTodoIds) => prevTodoIds
+          .filter((id) => id !== todoId));
+        setIsLoading(false);
+      });
+  };
+
+  const handleUpdateTodo = (todo: Todo, updatedProperties: Partial<Todo>) => {
+    setLoadingTodoIds((prevLoadingTodoIds) => [...prevLoadingTodoIds, todo.id]);
+
+    updateTodo({
+      ...todo,
+      ...updatedProperties,
+    })
+      .then((updatedTodo) => {
+        setTodos((prevTodos) => prevTodos.map((currentTodo) => (
+          currentTodo.id !== updatedTodo.id ? currentTodo : updatedTodo
+        )));
+      })
+      .catch(() => {
+        if (updatedProperties.title) {
+          setError(errorMessages.update_todo);
+        }
+      })
+      .finally(() => {
+        setLoadingTodoIds((prevTodoIds) => prevTodoIds
+          .filter((id) => id !== todo.id));
+        setIsLoading(false);
+      });
+  };
+
+  const toggleAllTodos = async () => {
+    setLoadingTodoIds(todos.map((todo) => todo.id));
+
+    try {
+      const areAllCompleted = todos.every((todo) => todo.completed);
+
+      const updatedTodos = await Promise.all(
```

Comment:
If not all completed update only todos that aren't completed.
It doesn't make sense to set the status `completed` for todos that are already completed